reference/protocol.ksy

```diff
@@ -296,8 +296,15 @@ types:
         size: 32
       - id: signature
         size: 64
-      - id: timestamp
+      - id: timestamp_and_vote_duration
         type: u8le
+    instances:
+        timestamp:
+          value: (timestamp_and_vote_duration >> 4)
```

Comment:
Using instances for the vote duration is the right idea.

However, the above would effectively divide the timestamp value by 16 in generated parsers, yielding the wrong value (you'd need to use a mask)

It looks like the node actually returns the timestamp with the duration bits intact (which is fine), so a separate timestamp instance is not needed.

As a result, I think `vote_common` should look like this:

```
  vote_common:
    doc: Common data shared by block votes and vote-by-hash votes
    seq:
      - id: account
        size: 32
      - id: signature
        size: 64
      - id: timestamp
        type: u8le
        doc: Number of seconds since the UTC epoch vote was generated at
    instances:
        vote_duration:
          value: (timestamp & 0xf)
          doc: Since V23.0 this is specified as 2^(duration + 4) in milliseconds

```

Reply:
Maybe the instance should be called `vote_duration_bits` though. The actual duration is `1 << (duration_bits + 4)`

Reply:
You are absolutely right, the shift is wrong. It shoudl either be left alone or it should be masked. Sorry @zhyatt , my bad.
(timestamp_and_vote_duration & 0xFFFFFFFFFFFFFFF0)

And vote_duration_bits sounds good.

Reply:
I think having an instance and zeroing the bottom 4 bits makes sense because it reflects the reality and is less random/jittery, even though it is probably harmless to not mask.

Reply:
Yeah, my vote would be to leave it unmasked as that accurately reflects the implementation we are describing.

Reply:
But unmasked it will include the vote_duration bits that is not part of the timestamp, which would be slightly inaccurate.
I am proposing this:
```
  vote_common:
    doc: Common data shared by block votes and vote-by-hash votes
    seq:
      - id: account
        size: 32
      - id: signature
        size: 64
      - id: timestamp_and_vote_duration
        type: u8le
    instances:
        timestamp:
          value: (timestamp_and_vote_duration & 0xfffffffffffffff0)
          doc: Number of seconds since the UTC epoch vote was generated at
        vote_duration:
          value: (timestamp_and_vote_duration & 0xf)
          doc: Since V23.0 this is specified as 2^(duration + 4) in milliseconds
```

Reply:
Yes, but _not_ masking seems to be what the node does for `timestamp` and the spec should reflect it IMO. That said, that's getting close to bikeshedding and I'm happy to accept both variations :)

Reply:
Hmmm, OK, I see your point now. The code is has half adopted the vote duration so either way is technically only half correct.

Reply:
@clemahieu what is your opinion on this?

Reply:
I see what you're saying about the timestamp value in vote, it's not masking off the low 4 bits though it should be since those are actually duration bits. Let's mask off the low 4 for timestamp and we'll update the node itself to do that also.

Reply:
Yes, I believe that is correct.

Reply:
I concur given the upcoming node change (maybe someone should make an issue for that so it's not lost?)

Reply:
Added the issue here https://github.com/nanocurrency/nano-node/issues/3705 and made commits to resolve this.